lib/lib.go

```diff
@@ -32,16 +45,417 @@ func init() {
+			}
+			path = filepath.Join(home, ".ipfs")
+		}
+
+		fsOpts := []ipfs.Option{
+			func(c *ipfs.StoreCfg) {
+				c.FsRepoPath = path
+			},
+			ipfs.OptsFromMap(cfg.Store.Options),
+		}
+		return ipfs.NewFilestore(fsOpts...)
+	case "map":
+		return cafs.NewMapstore(), nil
+	default:
+		return nil, fmt.Errorf("unknown store type: %s", cfg.Store.Type)
+	}
+}
+
+func newRegClient(cfg *config.Config) (rc *regclient.Client) {
+	if cfg.Registry != nil && cfg.Registry.Location != "" {
+		rc = regclient.NewClient(&regclient.Config{
+			Location: cfg.Registry.Location,
+		})
+	}
+	return
+}
+
+func newRepo(cfg *config.Config, store cafs.Filestore, rc *regclient.Client) (r repo.Repo, err error) {
+	var pro *profile.Profile
+	if pro, err = profile.NewProfile(cfg.Profile); err != nil {
+		return
+	}
+
+	switch cfg.Repo.Type {
+	case "fs":
+		return fsrepo.NewRepo(store, nil, pro, rc, cfg.Repo.Path)
+	case "mem":
+		return repo.NewMemRepo(pro, store, nil, profile.NewMemStore(), rc)
+	default:
+		return nil, fmt.Errorf("unknown repo type: %s", cfg.Repo.Type)
+	}
+}
+
+func newFilesystem(cfg *config.Config, store cafs.Filestore) (qfs.Filesystem, error) {
+	mux := map[string]qfs.PathResolver{
+		"local": localfs.NewFS(),
+		"http":  httpfs.NewFS(),
+		"cafs":  store,
 	}
+
+	if ipfss, ok := store.(*ipfs.Filestore); ok {
+		mux["ipfs"] = ipfss
+	}
+
+	fsys := muxfs.NewMux(mux)
+	return fsys, nil
+}
+
+// NewInstanceFromConfigAndNode is a temporary solution to create an instance from an
+// already-allocated QriNode & configuration
+// don't write new code that relies on this, instead create a configuration
+// and options that can be fed to NewInstance
+func NewInstanceFromConfigAndNode(cfg *config.Config, node *p2p.QriNode) Instance {
+	ctx, teardown := context.WithCancel(context.Background())
+	return &instance{
+		ctx:      ctx,
+		teardown: teardown,
+		cfg:      cfg,
+		node:     node,
+	}
+}
+
+// instance implements the (exported) Instance interface
+// create an instance one with NewInstance
+type instance struct {
+	ctx      context.Context
+	teardown context.CancelFunc
+
+	cfg *config.Config
+
+	streams  ioes.IOStreams
+	store    cafs.Filestore
+	qfs      qfs.Filesystem
+	registry *regclient.Client
+	repo     repo.Repo
+	node     *p2p.QriNode
+
+	rpc *rpc.Client
+}
+
+// Context returns the base context for this instance
+func (inst *instance) Context() context.Context {
+	return inst.ctx
+}
+
+// Config provides methods for manipulating Qri configuration
+func (inst *instance) Config() *config.Config {
+	return inst.cfg
+}
+
+// SetConfig implements the ConfigSetter interface
+func (inst *instance) SetConfig(cfg *config.Config) (err error) {
+
+	if path := inst.cfg.Path(); path != "" {
+		if err = cfg.WriteToFile(path); err != nil {
+			return
+		}
+	}
+
+	inst.cfg = cfg
+	return nil
+}
+
+// Node accesses the instance qri node if one exists
+func (inst *instance) Node() *p2p.QriNode {
+	return inst.node
```

Comment:
These bodies to me are the strongest evidence that we should use a non-interface `instance` struct in all of our internal code. There's no real alternative for how an interface could work aside from just returning these private fields. We should just be passing around a simple `instance` struct everywhere.